src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/AnonymousTypeManager.Templates.cs

```diff
@@ -163,60 +156,190 @@ private void CheckSourceLocationSeen(AnonymousTypePublicSymbol anonymous)
+                return typeArguments.Length == 0 ?
+                    template :
+                    template.Construct(typeArguments);
+            }
+            else
+            {
+                var typeParameters = GetReferencedTypeParameters(typeDescr);
+                var key = getTemplateKey(typeDescr, typeParameters);
+
+                // Get anonymous delegate template
+                AnonymousDelegateTemplateSymbol? template;
+                if (!this.SynthesizedDelegates.TryGetValue(key, out template))
+                {
+                    template = this.SynthesizedDelegates.GetOrAdd(key, new AnonymousDelegateTemplateSymbol(this, typeDescr, typeParameters));
+                }
+
+                // Adjust template location if the template is owned by this manager
+                if (ReferenceEquals(template.Manager, this))
+                {
+                    template.AdjustLocation(typeDescr.Location);
                 }
+
+                Debug.Assert(typeParameters.Length == template.TypeParameters.Length);
+                return typeParameters.Length == 0 ?
+                    template :
+                    template.Construct(typeParameters);
+            }
+
+            static bool allValidTypeArguments(AnonymousTypeDescriptor typeDescr)
+            {
+                var fields = typeDescr.Fields;
+                int n = fields.Length;
+                for (int i = 0; i < n - 1; i++)
+                {
+                    if (!isValidTypeArgument(fields[i]))
+                    {
+                        return true;
+                    }
+                }
+                var returnParameter = fields[n - 1];
+                return !returnParameter.Type.IsVoidType() && !isValidTypeArgument(returnParameter);
+            }
+
+            static bool isValidTypeArgument(AnonymousTypeField field)
+            {
+                return field.Type is { } type &&
+                    !type.IsPointerOrFunctionPointer() &&
+                    !type.IsRestrictedType();
+            }
+
+            static SynthesizedDelegateKey getTemplateKey(AnonymousTypeDescriptor typeDescr, ImmutableArray<TypeParameterSymbol> typeParameters)
+            {
+                if (typeParameters.Length > 0)
+                {
+                    var typeMap = new TypeMap(typeParameters, IndexedTypeParameterSymbol.Take(typeParameters.Length), allowAlpha: true);
+                    typeDescr = typeDescr.SubstituteTypes(typeMap, out bool changed);
+                    Debug.Assert(changed);
+                }
+                return new SynthesizedDelegateKey(typeDescr);
+            }
+        }
+
+        private static ImmutableArray<TypeParameterSymbol> GetReferencedTypeParameters(AnonymousTypeDescriptor typeDescr)
+        {
+            var referenced = PooledHashSet<TypeParameterSymbol>.GetInstance();
+            foreach (var field in typeDescr.Fields)
+            {
+                field.TypeWithAnnotations.VisitType(
```

Comment:
>VisitType [](http://example.com/codeflow?start=42&length=9)

I think we need to test scenarios with other anonymous types or anonymous delegates among these types, those can refer to type parameters as well. This means that we need the other part of the change to ```VisitType``` from that other PR. I'll ask to split that ```VisitType``` change into a separate PR so that we could merge that change and take advantage of it in this PR.

Reply:
Here is the PR for ```VisitType``` changes - [https://github.com/dotnet/roslyn/pull/58541](https://github.com/dotnet/roslyn/pull/58541).